corehq/apps/enterprise/enterprise.py

```diff
@@ -383,3 +388,68 @@ def _get_individual_export_rows(self, exports, export_line_counts):
             )
 
         return rows
+
+
+class EnterpriseCommCareVersionReport(EnterpriseReport):
+    title = _('CommCare Client Version Compliance')
+    metric = _('%% of mobile workers on the latest commcare client version ')
+
+    def __init__(self, account, couch_user, **kwargs):
+        super().__init__(account, couch_user, **kwargs)
+
+    @property
+    def headers(self):
+        return [
+            _('Mobile Worker'),
+            _('Project Space'),
+            _('Latest Version Available at Submission'),
+            _('Version in Use'),
+
+        ]
+
+    def rows_for_domain(self, domain_obj):
+        rows = []
+
+        for user in get_all_user_rows(domain_obj.name,
+                                    include_web_users=False,
+                                    include_mobile_users=True,
+                                    include_inactive=False,
+                                    include_docs=True):
+            user = CouchUser.wrap_correctly(user['doc'])
+            last_submission = user.reporting_metadata.last_submission_for_user
+            last_used_device = user.last_device
+
+            version_in_use = _('Unknown')
+            date_of_use = None
+
+            # If the user hasn't submitted a form, we use the last used device
+            if last_submission and last_submission.commcare_version:
+                version_in_use = _format_commcare_version(last_submission.commcare_version)
+                date_of_use = last_submission.submission_date
+            elif last_used_device and last_used_device.commcare_version:
+                version_in_use = _format_commcare_version(last_used_device['commcare_version'])
+                date_of_use = last_used_device.last_used
+
+            latest_version_at_time_of_use = get_latest_version_at_time(date_of_use)
+
+            if is_out_of_date(version_in_use, latest_version_at_time_of_use):
+                rows.append([
+                    user.username,
+                    domain_obj.name,
+                    latest_version_at_time_of_use,
+                    version_in_use,
+                ])
+
+        return rows
+
+    @property
+    def total(self):
+        total_mobile_workers = 0
+        total_up_to_date = 0
+        for domain_obj in self.domains():
+            domain_mobile_workers = get_mobile_user_count(domain_obj.name, include_inactive=False)
+            if domain_mobile_workers:
+                total_mobile_workers += domain_mobile_workers
+                total_up_to_date += domain_mobile_workers - len(self.rows_for_domain(domain_obj))
```

Comment:
There is definitely a performance concern here, as if this summary query takes 2 minutes, that is 2 minutes blocking a web worker. It's worth trying to see if Elasticsearch is any faster here. Otherwise, we may want to cache this statistic.

Reply:
1700c2f6d664f5435b938324e2ca24365ac560fc

Reply:
Tried on production again, it takes between 2 minutes and 3 minutes. Last time I was about to leave my apartment so I didn't wait until it finished, so I guess last time when I query couch directly maybe it longer than 3 minutes. There is improvement by using elastic search. Is the improvement acceptable? I think maybe yes? Because not every enterprise account would have that amount of mobile workers...

Reply:
If we're talking minute I think we need to think about caching or turning this into a celery task. I can imagine that multiple requests to this page could cause a disruption in our responsiveness that would be hard to diagnose